lib/internal/fs/promises.js

```diff
@@ -790,50 +782,56 @@ async function readdirRecursive(originalPath, options) {
         !!options.withFileTypes,
         kUsePromises,
       ),
+      0,
     ],
   ];
 
-
   if (options.withFileTypes) {
     while (queue.length > 0) {
       // If we want to implement BFS make this a `shift` call instead of `pop`
-      const { 0: path, 1: readdir } = ArrayPrototypePop(queue);
+      const { 0: path, 1: readdir, 2: depth } = ArrayPrototypePop(queue);
       for (const dirent of getDirents(path, readdir)) {
-        ArrayPrototypePush(result, dirent);
-        if (dirent.isDirectory()) {
-          const direntPath = pathModule.join(path, dirent.name);
-          ArrayPrototypePush(queue, [
-            direntPath,
-            await binding.readdir(
+        const direntPath = pathModule.join(path, dirent.name);
+        if (typeof options.filter !== 'function' || options.filter(direntPath, depth)) {
+          ArrayPrototypePush(result, dirent);
+          if (dirent.isDirectory()) {
+            ArrayPrototypePush(queue, [
               direntPath,
-              options.encoding,
-              true,
-              kUsePromises,
-            ),
-          ]);
+              await binding.readdir(
+                direntPath,
+                options.encoding,
+                true,
+                kUsePromises,
+              ),
+              depth + 1,
+            ]);
+          }
         }
       }
     }
   } else {
     while (queue.length > 0) {
-      const { 0: path, 1: readdir } = ArrayPrototypePop(queue);
+      const { 0: path, 1: readdir, 2: depth } = ArrayPrototypePop(queue);
       for (const ent of readdir) {
         const direntPath = pathModule.join(path, ent);
-        const stat = binding.internalModuleStat(direntPath);
-        ArrayPrototypePush(
-          result,
-          pathModule.relative(originalPath, direntPath),
-        );
-        if (stat === 1) {
-          ArrayPrototypePush(queue, [
-            direntPath,
-            await binding.readdir(
-              pathModule.toNamespacedPath(direntPath),
-              options.encoding,
-              false,
-              kUsePromises,
-            ),
-          ]);
+        if (typeof options.filter !== 'function' || options.filter(direntPath, depth)) {
```

Comment:
```suggestion
        if (typeof options.filter === 'function' && !options.filter(direntPath, depth))
          continue;
```
If I'm reading this part correctly, we can invert this `if` to reduce nesting.